avocado/core/task/runtime.py

```diff
@@ -114,35 +116,45 @@ def get_dependencies_form_runnable(cls, runnable, status_server_uri=None,
         """
-        if runnable.dependencies is None:
-            return []
+
+        def _crete_runtime_tasks(dependencies):
+            # creates the tasks and runtime tasks for the dependencies
+            dependencies_runtime_tasks = []
+            for dependency_runnable in dependencies:
+                name = '%s-%s' % (dependency_runnable.kind,
+                                  dependency_runnable.kwargs.get('name'))
+                prefix = 0
+                # the human UI works with TestID objects, so we need to
+                # use it to name Task
+                task_id = TestID(prefix, name)
+                # with --dry-run we don't want to run dependencies
+                if runnable.kind == 'dry-run':
+                    dependency_runnable.kind = 'noop'
+                # creates the dependency task
+                dependency_task = Task(dependency_runnable,
+                                       identifier=task_id,
+                                       status_uris=status_server_uri,
+                                       category='dependency',
+                                       job_id=job_id)
+                dependencies_runtime_tasks.append(cls(dependency_task))
+            return dependencies_runtime_tasks
+
+        pre_runnables = list(chain.from_iterable(
+            TestPreDispatcher().map_method_with_return('pre_test_runnables',
+                                                       runnable)))
+        post_runnables = list(chain.from_iterable(
+            TestPostDispatcher().map_method_with_return('post_test_runnables',
+                                                        runnable)))
 
         # creates the runnables for the dependencies
-        dependencies_runnables = DependencyResolver.resolve(runnable)
-        dependencies_runtime_tasks = []
-        # creates the tasks and runtime tasks for the dependencies
-        for dependency_runnable in dependencies_runnables:
-            name = '%s-%s' % (dependency_runnable.kind,
-                              dependency_runnable.kwargs.get('name'))
-            prefix = 0
-            # the human UI works with TestID objects, so we need to
-            # use it to name Task
-            task_id = TestID(prefix, name)
-            # with --dry-run we don't want to run dependencies
-            if runnable.kind == 'dry-run':
-                dependency_runnable.kind = 'noop'
-            # creates the dependency task
-            dependency_task = Task(dependency_runnable,
-                                   identifier=task_id,
-                                   status_uris=status_server_uri,
-                                   category='dependency',
-                                   job_id=job_id)
-            dependencies_runtime_tasks.append(cls(dependency_task))
-
-        return dependencies_runtime_tasks
+        if runnable.dependencies:
+            pre_runnables = pre_runnables + DependencyResolver.resolve(runnable)
```

Comment:
This is an interesting decision which needs to be either better discussed or documented: should the pre-runnables be executed before or after the dependencies?

On a first quick thought, I agree with this (pre-runnables before the dependencies), but suppose we're using both stock features:
* pre-test sysinfo
* package depdendencies

If we need information that a test passed or failed with/without a given package, this may *not* be obtained from sysinfo.

Reply:
I decided to remove the `DependencyResolver` and use only pre-runnables for it in 6afb2e2407b472a7845e48aaa7e82803530f7ec2. But I haven't thought about order of pre-test task yet, and you are right that it is something which have to be solved. Right now pre-test tasks are on the same level, and they will be run simultaneously in parallel run base on the state machine behavior. I agree with you that this is not an optimal solution.

Reply:
I think we can solve this problem with the solution from #4939 or maybe just use the same behavior as `plugins.resolver.order` has.